src/Microsoft.NetCore.Analyzers/UnitTests/Security/DoNotUseInsecureDeserializerBinaryFormatterWithoutBinderTests.cs

```diff
@@ -0,0 +1,602 @@
+        {
+            string myBinderCSharpSourceCode = @"
+using System;
+using System.Runtime.Serialization;
+
+namespace Blah
+{
+    public class MyBinder : SerializationBinder
+    {
+        public override Type BindToType(string assemblyName, string typeName)
+        {
+            throw new NotImplementedException();
+        }
+    }
+}
+            ";
+
+            this.VerifyCSharp(
+                new[] { source, myBinderCSharpSourceCode }.ToFileAndSource(), 
+                expected);
+        }
+
+        [Fact]
+        public void Deserialize_Diagnostic()
+        {
+            VerifyCSharp(@"
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Blah
+{
+    public class Program
+    {
+        public object BfUnsafeDeserialize(byte[] bytes)
+        {
+            BinaryFormatter formatter = new BinaryFormatter();
+            return formatter.Deserialize(new MemoryStream(bytes));
+        }
+    }
+}",
+            GetCSharpResultAt(12, 20, BinderNotSetRule, "object BinaryFormatter.Deserialize(Stream serializationStream)"));
+        }
+
+        [Fact]
+        public void Deserialize_OptionalParameters_Diagnostic()
+        {
+            VerifyCSharp(@"
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Blah
+{
+    public class Program
+    {
+        public object BfDeserialize(byte[] bytes = null)
+        {
+            BinaryFormatter formatter = new BinaryFormatter();
+            return formatter.Deserialize(new MemoryStream(bytes));
+        }
+    }
+}",
+            GetCSharpResultAt(12, 20, BinderNotSetRule, "object BinaryFormatter.Deserialize(Stream serializationStream)"));
+        }
+
+        // Ideally, we'd detect that formatter.Binder is always null.
+        [Fact]
+        public void DeserializeWithInstanceField_Diagnostic_NotIdeal()
+        {
+            VerifyCSharp(@"
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Blah
+{
+    public class Program
+    {
+        BinaryFormatter formatter = new BinaryFormatter();
+
+        public object BfUnsafeDeserialize(byte[] bytes)
+        {
+            return this.formatter.Deserialize(new MemoryStream(bytes));
+        }
+    }
+}",
+            GetCSharpResultAt(13, 20, BinderMaybeNotSetRule, "object BinaryFormatter.Deserialize(Stream serializationStream)"));
+        }
+
+        [Fact(Skip = "Ideally, we'd detect that this.formatter.Binder is set.")]
+        public void DeserializeWithInstanceField_NoDiagnostic()
+        {
+            VerifyCSharpWithMyBinderDefined(@"
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Blah
+{
+    public class Program
+    {
+        BinaryFormatter formatter = new BinaryFormatter() { Binder = new MyBinder() };
+
+        public object BfUnsafeDeserialize(byte[] bytes)
+        {
+            return this.formatter.Deserialize(new MemoryStream(bytes));
```

Comment:
>formatter [](start = 24, length = 9)

Tricky part is you don't know if something else read the field, passed it around to some other assembly and cleared the binder before `BfUnsafeDeserialize` is called.

Reply:
`BinaryFormatter.Binder` seems to have a public setter.

---
In reply to: [224093203](https://github.com/dotnet/roslyn-analyzers/pull/1838#discussion_r224093203) [](ancestors = 224093203)

Reply:
Aye, that's what the "maybe it's set maybe it's not" rule is for.